src/Controller/BackendController.php

```diff
@@ -238,4 +242,47 @@ public function pickerAction(Request $request): RedirectResponse
 
         return new RedirectResponse($picker->getCurrentUrl());
     }
+
+    /**
+     * @return Response
+     *
+     * @Route("/contao/2fa", name="contao_backend_2fa")
+     */
+    public function twoFactorAuthAction(): Response
+    {
+        $this->get('contao.framework')->initialize();
+
+        /** @var RouterInterface $router */
+        $router = $this->get('router');
+        $token = $this->get('security.token_storage')->getToken();
+
+        if (!$token instanceof TwoFactorToken) {
+            return new RedirectResponse($router->generate('contao_backend_login'));
+        }
+
+        $authenticatedToken = $token->getAuthenticatedToken();
+
+        if (!$authenticatedToken instanceof UsernamePasswordToken) {
+            return new RedirectResponse($router->generate('contao_backend_login'));
+        }
+
+        $user = $authenticatedToken->getUser();
+
+        if (!$user instanceof BackendUser) {
+            return new RedirectResponse($router->generate('contao_backend_login'));
+        }
+
+        return $this->loginAction();
+    }
+
+    /**
+     * Dummy action for Symfony security.
```

Comment:
Are we 100% sure that this dummy action is required? @aschempp /cc

Reply:
It's required because of the two-factor bundle. See https://github.com/scheb/two-factor-bundle/pull/145, once it gets fixed we can remove this.